src/adapters/e4c-ranger-staking.adapter.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright Abridged, Inc. 2022. All Rights Reserved.
+// Node module: @collabland/staking-contracts
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {BindingScope, extensionFor, injectable} from '@loopback/core';
+import {BigNumber} from 'ethers';
+import {STAKING_ADAPTERS_EXTENSION_POINT} from '../keys';
+import {BaseStakingContractAdapter, StakingAsset} from '../staking';
+import {E4cRangerStaking__factory} from '../types/factories/E4cRangerStaking__factory';
+
+@injectable(
+  {
+    scope: BindingScope.SINGLETON,
+  },
+  extensionFor(STAKING_ADAPTERS_EXTENSION_POINT),
+)
+export class E4CRangerStakingContractAdapter extends BaseStakingContractAdapter {
+  contractAddress = '0x9c18beA91AE053397918410311dbB89295baE18b';
+
+  supportedAssets: StakingAsset[] = [
+    {
+      asset: 'ERC721:0xbA265B93519E6473F34F46ee35F4B23970F41a3f',
+    },
+  ];
+
+  async getStakedTokenIds(owner: string): Promise<BigNumber[]> {
+    const contract = E4cRangerStaking__factory.connect(
+      this.contractAddress,
+      this.provider,
+    );
+
+    const z = [];
+
+    for (let i = 0; i < 1024; ++i) {
```

Comment:
Hmm, it will make 1024 calls to the staking contract. Is there a way to optimize this? At least we need to invoke `originalOwner()` concurrently using a utility such as https://www.npmjs.com/package/p-map.

Reply:
I am afraid there is no simple way to do so. Since the contract was deployed much early before we found this repo.

Reply:
We probably need to use https://github.com/Destiner/ethcall.

Reply:
I used the following code in a similar case:

```ts
 
  /**
   * Invoke multiple on-chain calls in a batch
   * @param calls - Contract calls
   * @param options
   * @returns
   */
  private async multicall<T>(calls: Call[], options?: PMapByPageOptions) {
    debug('Multicalls: %O', calls);
    if (this.multicallProvider == null) {
      this.multicallProvider = new MulticallProvider();
      await this.multicallProvider.init(this.provider);
    }

    const result = await pMapByPage(
      calls,
      page => this.multicallProvider.tryAll<T>(page),
      {pageSize: 100, ...options},
    );
    const response = result.flat();
    debug('Response: %O', response);
    return response;
  }
```

Reply:
Done.